File: zig/words/src/numerical.zig
pub fn fibonacci(n: i32) i32 {
    if (n < 0) {
        return -1;
    } else if (n < 2) {
        return n;
    }
    return fibonacci_calc(n, 0, 1);
}

fn fibonacci_calc(n: i32, acc2: i32, acc1: i32) i32 {
    if (n == 2) {
        return acc1 + acc2;
    }
    return fibonacci_calc(n - 1, acc1, acc1 + acc2);
}

pub fn factorial(n: i32) i32 {
    if (n < 0) {
        return -1;
    }
    return factorial_calc(n, 1);
}

fn factorial_calc(n: i32, acc: i32) i32 {
    if (n <= 1) {
        return acc;
    }
    return factorial_calc(n - 1, n * acc);
}

pub fn gcd(a: i32, b: i32) i32 {
    if (b == 0) {
        return a;
    }
    return gcd(b, @mod(a, b));
}

pub fn lcm(a: i32, b: i32) i32 {
    return @divExact(a, gcd(a, b)) * b;
}

File: nim/words/src/numerical.nim
proc fibonacci*(n: int): int =
  if n < 0: return -1
  elif n < 2: return n
  else:
    var acc1 = 1
    var acc2 = 0
    for _ in 2..<n:
      (acc2, acc1) = (acc1, acc1 + acc2)
    return acc1 + acc2

proc factorial*(n: int): int =
  if n < 0: return -1
  else:
    var acc = 1
    for i in 2..n:
      acc *= i
    return acc

proc gcd*(a, b: int): int =
  var x = a
  var y = b
  while y != 0:
    (x, y) = (y, x mod y)
  return x

proc lcm*(a, b: int): int =
  return (a div gcd(a, b)) * b